src/mame/drivers/firebeat.cpp

```diff
@@ -154,14 +155,213 @@
+	auto order = WDL_fft_permute_tab(FFT_LENGTH / 2);
+	for (int ch = 0; ch < TOTAL_CHANNELS; ch++) {
+		double notch_max[TOTAL_BARS] = { -1, -1, -1, -1, -1, -1 };
+		int cur_notch = 0;
+
+		for (int i = 0; i <= FFT_LENGTH / 2; i++) {
+			const double freq = (double)i / FFT_LENGTH * srate;
+
+			if (freq < NOTCHES[cur_notch]) {
+				continue;
+			}
+
+			if (freq > NOTCHES[cur_notch+1]) {
+				cur_notch++;
+			}
+
+			if (cur_notch >= LAST_NOTCH) {
+				// Don't need to calculate anything above this frequency
+				break;
+			}
+
+			WDL_FFT_COMPLEX* bin = (WDL_FFT_COMPLEX*)m_fft_buf[ch] + order[i];
+
+			const double re = bin->re;
+			const double im = bin->im;
+			const double mag = sqrt(re*re + im*im);
+
+			if (notch_max[cur_notch] == -1 && freq >= NOTCHES[cur_notch] && freq < NOTCHES[cur_notch+1]) {
+				notch_max[cur_notch] = mag;
+			}
+		}
+
+		for (int i = 0; i < TOTAL_BARS; i++) {
+			double val = log10(notch_max[i] * 4096) * 20;
+			val = std::max<double>(0, val);
+			m_bars[ch][i] = uint32_t(std::min<double>(val, 255.0f));
+		}
+	}
+}
+
+void firebeat_extend_spectrum_analyzer_device::apply_fft(uint32_t buf_index)
+{
+	float *audio_l = m_audio_buf[buf_index][0];
+	float *audio_r = m_audio_buf[buf_index][1];
+	float *buf_l = m_fft_buf[0];
+	float *buf_r = m_fft_buf[1];
+
+	for (int i = 0; i < FFT_LENGTH; i++)
+	{
+		*buf_l++ = *audio_l++;
+		*buf_r++ = *audio_r++;
+	}
+
+	for (int ch = 0; ch < TOTAL_CHANNELS; ch++) {
+		WDL_real_fft((WDL_FFT_REAL*)m_fft_buf[ch], FFT_LENGTH, 0);
+
+		for (int i = 0; i < FFT_LENGTH; i++) {
+			m_fft_buf[ch][i] /= (WDL_FFT_REAL)FFT_LENGTH;
+		}
+	}
+}
+
+void firebeat_extend_spectrum_analyzer_device::update_fft()
+{
+	m_audio_count[m_audio_fill_index]++;
+	if (m_audio_count[m_audio_fill_index] >= FFT_LENGTH)
+	{
+		apply_fft(m_audio_fill_index);
+
+		m_audio_fill_index = 1 - m_audio_fill_index;
+		m_audio_count[m_audio_fill_index] = 0;
+	}
+}
+
+uint8_t firebeat_extend_spectrum_analyzer_device::read(offs_t offset)
+{
+	// Visible in the sound test menu and used for the spectral analyzer game skin
+	// The actual data is coming from a circuit built on the extend board made up of NJU7507 x2 for each channel of audio.
+	// The spectrum analyzer circuit is only populated on the extend board used by beatmania III but the footprints and
+	// labels are still on the PCB for other games that use the extend board.
+
+	// Return values notes:
+	// - The values are based on the results of a chained NJU7507 which should give 14 bands worth of frequency data. However, only 6 bands are used in-game.
+	// - Anything <= 8 will not display anything in-game and seems unused in memory
+	// - It reads the upper and lower half of the register separately as bytes, but it the upper byte doesn't seem like it's actually used.
+	// - In-game the skin shows up to +9 dB but it actually caps out at -3 dB on the skin
+
+	int ch = offset >= 0x40; // 0 = Left, 1 = Right
+	int notch = 6 - (((offset >> 2) & 0x0f) >> 1);
```

Comment:
Wouldn’t this be equivalent to `6 - ((offset >> 3) & 7)` or am I missing something?